### PR TITLE
Support Cairo 0.10

### LINKED ISSUE #170: Remove makefile from `nile init` standard project

### LINKED ISSUE #182: Migrate nile to Cairo 0.10

### PR DESCRIPTION
**Pull request by martriay**
Fixes #170 and #180 and #182. Supercedes #181.
This PR also drops support for python 3.7 as both cairo-lang and starknet-devnet did.

### CODE REVIEW

File: src/nile/core/account.py [hunk 1]
```diff
@@ -68,32 +69,25 @@ def send(self, to, method, calldata, max_fee, nonce=None):
...
             max_fee=max_fee,
         )
 
-        params = []
-        params.append(len(call_array))
-        params.extend(*call_array)
-        params.append(len(calldata))
-        params.extend(calldata)
-        params.append(nonce)
-
```
Comment (andrew-fleming) [inline src/nile/core/account.py:91]: Yes!!! :)

**Review by andrew-fleming — COMMENTED** [review]
Looks great, @martriay! I left a question and I'm wondering if we should add `declare` to the `Account` class since declaration txs should only be called from accounts (see [https://github.com/starkware-libs/cairo-lang/releases](https://github.com/starkware-libs/cairo-lang/releases)) What do you think?

**Comment by andrew-fleming** [comment]
Also, is `tox` working on your local machine? I'm getting this error: `ERROR:   default: could not install deps [cairo-lang==0.10.0, starknet-devnet==0.3.0, marshmallow-dataclass==8.5.3];`

I tried pinning devnet to its latest release, but it only works if I don't list it as a dep

File: tests/resources/contracts/account_mock.cairo [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
-# if found, it expects an "--account_contract" flag during compilation
+// The StarkNet compiler checks for an external "__execute__" method
+// if found, it expects an "--account_contract" flag during compilation
 @external
-func __execute__():
-    return ()
-end
+func __execute__() {
+    return ();
+}
```
[the remark this answers is not in the document]
Comment (ericnordelo) [reply tests/resources/contracts/account_mock.cairo:8]: Is this mock used in the project? I can't find where.

[the remark this answers is not in the document]
Comment (ericnordelo) [reply tests/resources/contracts/account_mock.cairo:8]: I'm curious too. Maybe the compiler is going to yell in version 0.11 but for now is still accepting "transaction 0 format" accounts (deprecated but no removed) ? Or does the compiler allow entrypoint external functions even when the interface is not completed (not recognizing the smart contract as an Account but allowing the deployment) ?

**Review by ericnordelo — COMMENTED** [review]
I left a couple of comments but in general LGTM.

File: tests/resources/contracts/account_mock.cairo (hunk 2 above)
[the remark this answers is not in the document]
Comment (martriay) [reply tests/resources/contracts/account_mock.cairo:8]: actually this file is not being used, i'll remove it

[the remark this answers is not in the document]
Comment (martriay) [reply tests/resources/contracts/account_mock.cairo:8]: it does complain if you try to compile a contract without it:
```
🔨 Compiling src/openzeppelin/account/presets/Account.cairo
Account contracts must have external functions named {'__execute__', '__validate_declare__', '__validate__'}, found: ['__execute__', '__validate__'].
```

**Comment by martriay** [comment]
I'm not sure about changing `deploy` and `declare` yet. Mainly because our account [doesn't implement niether](https://github.com/OpenZeppelin/cairo-contracts/issues/384#issuecomment-1198159762) of both and we don't have a [deployer contract](https://community.starknet.io/t/universal-deployer-contract-proposal/1864) yet. Therefore, if I were to change `deploy` and `declare` to go through an Account, it's not clear what function should I be calling.

With the current implementation, users will still be able to call such functions with the `send` method. That  being said, I had no issue declaring and deploying contracts on goerli with the current `deploy` and `declare` methods. I reported it though because supposedly it shouldn't be possible with 0.10.

**Comment by martriay** [comment]
> I'm wondering if we should add `declare` to the `Account` class since declaration txs should only be called from accounts

I'm not sure about changing `deploy` and `declare` yet. Mainly because our account [doesn't implement niether](https://github.com/OpenZeppelin/cairo-contracts/issues/384#issuecomment-1198159762) of both and we don't have a [deployer contract](https://community.starknet.io/t/universal-deployer-contract-proposal/1864) yet. Therefore, if I were to change `deploy` and `declare` to go through an Account, it's not clear what function should I be calling.

With the current implementation, users will still be able to call such functions with the `send` method. That  being said, I had no issue declaring and deploying contracts on goerli with the current `deploy` and `declare` methods. I reported it though because supposedly it shouldn't be possible with 0.10.

Thoughts?



> Also, is `tox` working on your local machine? I'm getting this error: `ERROR: default: could not install deps [cairo-lang==0.10.0, starknet-devnet==0.3.0, marshmallow-dataclass==8.5.3];`
> 
> I tried pinning devnet to its latest release, but it only works if I don't list it as a dep

I pinned the latest now, but yes it works perfectly on my machine. @ericnordelo what about you?

**Comment by ericnordelo** [comment]
> I pinned the latest now, but yes it works perfectly on my machine. @ericnordelo what about you?

Works well for me.

**Comment by andrew-fleming** [comment]
> I'm not sure about changing deploy and declare yet. Mainly because our account https://github.com/OpenZeppelin/cairo-contracts/issues/384#issuecomment-1198159762 of both and we don't have a [deployer contract](https://community.starknet.io/t/universal-deployer-contract-proposal/1864) yet. Therefore, if I were to change deploy and declare to go through an Account, it's not clear what function should I be calling.

Great points! 

> With the current implementation, users will still be able to call such functions with the send method. That being said, I had no issue declaring and deploying contracts on goerli with the current deploy and declare methods. I reported it though because supposedly it shouldn't be possible with 0.10.

Hmm I imagine they're allowing a transitionary period for v0.10 in line with supporting tx version `0`. Since there's still leeway, I agree in waiting. The more time to discuss and plan while anticipating the forthcoming changes, the better :)

**Review by andrew-fleming — APPROVED** [review]
Great work! Looks good to go :)

**Comment by martriay** [comment]
> Hmm I imagine they're allowing a transitionary period for v0.10 in line with supporting tx version `0`.

That would make sense: declaration and deployment are using tx `0` and then I use `1` with the account.

**Pull request merged**